app/src/main/java/com/minimalisticapps/priceconverter/presentation/home/viewmodels/HomeViewModel.kt

```diff
@@ -136,32 +131,114 @@ class HomeViewModel @Inject constructor(
     //    function for getting fiat coins
     fun getFiatCoins() {
         viewModelScope.launch {
-            getFiatCoinUseCase().collect {
-                var position = -1
-                _fiatCoinsList.value = it.map { bitPayCoinWithFiatCoin ->
-                    Pair(position++, bitPayCoinWithFiatCoin)
+            getFiatCoinUseCase().collect { it ->
+                val listStateValue: MutableList<Pair<Int, BitPayCoinWithFiatCoin>> = mutableListOf()
+                val shitcoinInputsStateValue: MutableMap<String, MutableState<TextFieldValue>> =
+                    mutableMapOf()
+
+                it.forEachIndexed { index, shitcoin ->
+                    listStateValue.add(Pair(index, shitcoin))
+
+                    val code = shitcoin.fiatCoinExchange.code
+                    shitcoinInputsStateValue[code] = shitcoinInputsState[code]
+                        ?: mutableStateOf(TextFieldValue())
+                }
+
+                _shitcoinListState.value = listStateValue
+                shitcoinInputsState = shitcoinInputsStateValue
+
+                if (selectedCoin.value == "BTC") {
+                    recalculateByBitcoin()
+                } else {
+                    val index =
+                        _shitcoinListState.value.find { it2 -> it2.second.fiatCoinExchange.code == selectedCoin.value }?.first
+                    if (index != null) {
+                        recalculateByShitcoin(index)
+                    }
                 }
             }
 
         }
     }
 
-    fun deleteFiatCoin(fiatCoinExchange: FiatCoinExchange) {
+    fun deleteFiatCoin(index: Int?) {
+        val shitcoin = _shitcoinListState.value.find { it.first == index } ?: return
+
         viewModelScope.launch {
-            deleteUseCase.invoke(fiatCoinExchange)
+            deleteUseCase.invoke(shitcoin.second.fiatCoinExchange)
         }
     }
 
-    fun setTextFieldValueBtc(text: String) {
-        val formatted = formatBtc(text.toBigDecimal())
-        _textFiledValueBtc.value = TextFieldValue(
-            formatted,
-            TextRange(formatted.length)
-        )
+    fun updateBitcoinAmount(value: TextFieldValue) {
+        // if only selection changed don't do any magic, just change selection
+        // do NOT recalculate, it would cumulate rounding error
+        if (textFiledValueBtc.value.text == value.text) {
+            textFiledValueBtc.value = value
+            return
+        }
+
+        textFiledValueBtc.value =
+            updateTextFieldModelWithCommas(textFiledValueBtc.value, value) { formatBtcString(it) }
+
+        recalculateByBitcoin()
     }
 
-    fun setTextFieldValueBtc(textFieldValue: TextFieldValue) {
-        _textFiledValueBtc.value = textFieldValue
+    private fun updateBitcoinAmountWithoutRecalculate(value: BigDecimal) {
+        textFiledValueBtc.value = TextFieldValue(text = formatBtc(value))
+    }
+
+    private fun recalculateByShitcoin(shitcoinIndex: Int) {
```

Comment:
The logic of re-calculation was implemented in the component. This is absolutely BAD so I have at leated moved it into model into one place - the recalculate method.